File: hydra-web/src/api/commands.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from '@/lib/api-client';
import { queryKeys } from '@/lib/query-client';
import type { ApiResponse } from '@/types/api';

export type CommandType =
  | 'metadata'
  | 'service'
  | 'package'
  | 'config'
  | 'system'
  | 'custom'
  | 'node'
  | 'agent';
export type CommandCategory = 'service' | 'node' | 'agent';
export type CommandStatus =
  | 'pending'
  | 'rejected'
  | 'queued'
  | 'executing'
  | 'completed'
  | 'failed'
  | 'timeout'
  | 'cancelled';
export type CommandExecutionMethod = 'agent-direct' | 'agent-poll' | 'integration';
export type CommandDeliveryMode = 'direct_or_poll' | 'poll_only';

export interface CommandTarget {
  nodeId: string;
  serviceId?: string | null;
}

export interface CreateCommandRequest {
  registryId: string;
  target: CommandTarget;
  parameters?: Record<string, unknown> | null;
  timeoutSeconds?: number;
}

export interface CommandResult {
  success: boolean;
  output?: string | null;
  exitCode?: number | null;
  error?: string | null;
  data?: Record<string, unknown> | null;
}

export interface CommandError {
  code: string;
  message: string;
  details?: Record<string, unknown> | null;
}

export interface RequestedBy {
  userId?: string | null;
  source: string;
  clientId?: string | null;
}

export interface CommandQueuedResponse {
  commandId: string;
  registryId?: string;
  type: CommandType;
  target: CommandTarget;
  action: string;
  status: CommandStatus;
  executionMethod: CommandExecutionMethod;
  result?: CommandResult | null;
  queuePosition?: number | null;
  queuedAt?: string | null;
  completedAt?: string | null;
}

export interface CommandSummary {
  commandId: string;
  registryId?: string | null;
  type: CommandType;
  target: CommandTarget;
  action: string;
  status: CommandStatus;
  createdAt: string;
}

export interface CommandResponse {
  commandId: string;
  registryId?: string | null;
  type: CommandType;
  target: CommandTarget;
  action: string;
  parameters?: Record<string, unknown> | null;
  status: CommandStatus;
  executionMethod?: CommandExecutionMethod | null;
  result?: CommandResult | null;
  error?: CommandError | null;
  requestedBy?: RequestedBy | null;
  timeoutSeconds: number;
  retryCount: number;
  queuePosition?: number | null;
  chain?: Record<string, unknown> | null;
  createdAt: string;
  queuedAt?: string | null;
  startedAt?: string | null;
  completedAt?: string | null;
  cancelledAt?: string | null;
  cancelledBy?: string | null;
}

export interface CommandDefinitionSummary {
  registryId: string;
  category: CommandCategory;
  action: string;
  displayName: string;
  description?: string | null;
  minimumRole: string;
  requiresConfirmation: boolean;
  timeout: number;
  deliveryMode: CommandDeliveryMode;
  builtIn: boolean;
  deprecated: boolean;
}

export interface QueueStats {
  totalQueued: number;
  totalExecuting: number;
  oldestQueuedAt?: string | null;
}

export interface QueueViewResponse {
  queue: CommandSummary[];
  stats: QueueStats;
}

export interface CommandCancelledResponse {
  commandId: string;
  status: CommandStatus;
  cancelledAt: string;
  cancelledBy?: string | null;
}

// ── Mutations ────────────────────────────────────────────────────────────

export function useCreateCommand() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (data: CreateCommandRequest) => {
      const response = await apiClient.post<ApiResponse<CommandQueuedResponse>>(
        '/commands',
        data
      );
      return response.data.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: queryKeys.commands.list() });
      queryClient.invalidateQueries({ queryKey: queryKeys.commands.queue() });
    },
  });
}

export function useCancelCommand() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (commandId: string) => {
      const response = await apiClient.post<ApiResponse<CommandCancelledResponse>>(
        `/commands/${commandId}/cancel`
      );
      return response.data.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: queryKeys.commands.all });
    },
  });
}

export function useFlushQueue() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (scope: string = 'all') => {
      const response = await apiClient.post<ApiResponse<{ flushedCount: number }>>(
        '/commands/queue/flush',
        { scope, confirm: true }
      );
      return response.data.data;
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: queryKeys.commands.all });
    },
  });
}

// ── Queries ──────────────────────────────────────────────────────────────

export function useCommands(params?: { nodeId?: string; status?: CommandStatus; type?: CommandType; registryId?: string }) {
  return useQuery({
    queryKey: queryKeys.commands.list(params),
    queryFn: async () => {
      const response = await apiClient.get<ApiResponse<CommandSummary[]>>('/commands', {
        params,
      });
      return response.data.data;
    },
  });
}

export function useCommand(commandId: string | undefined) {
  return useQuery({
    queryKey: queryKeys.commands.detail(commandId ?? ''),
    queryFn: async () => {
      const response = await apiClient.get<ApiResponse<CommandResponse>>(
        `/commands/${commandId}`
      );
      return response.data.data;
    },
    enabled: !!commandId,
    refetchInterval: (query) => {
      const data = query.state.data;
      if (data && (data.status === 'queued' || data.status === 'executing')) {
        return 2000; // Poll every 2s while active
      }
      return false;
    },
  });
}

export function useCommandQueue(params?: { nodeId?: string }) {
  return useQuery({
    queryKey: queryKeys.commands.queue(params),
    queryFn: async () => {
      const response = await apiClient.get<ApiResponse<QueueViewResponse>>('/commands/queue', {
        params,
      });
      return response.data.data;
    },
    refetchInterval: 5000, // Auto-refresh every 5s
  });
}

export function useCommandCatalog(params?: { category?: CommandCategory; includeDeprecated?: boolean }) {
  return useQuery({
    queryKey: queryKeys.commandCatalog.list(params),
    queryFn: async () => {
      const response = await apiClient.get<ApiResponse<CommandDefinitionSummary[]>>('/command-catalog', {
        params,
      });
      return response.data.data;
    },
  });
}

File: hydra-api/hydra/api/v1/models/commands.py
"""Command models for command execution queue and history."""

from datetime import datetime
from enum import Enum
from typing import Annotated

from pydantic import BaseModel, Field, field_validator

from hydra.api.v1.core.validators import (
    NODE_ID_PATTERN_NEW,
    SERVICE_ID_PATTERN,
    validate_node_id_strict,
    validate_service_id,
)


class CommandType(str, Enum):
    """Types of commands that can be executed."""

    METADATA = "metadata"
    SERVICE = "service"
    PACKAGE = "package"
    CONFIG = "config"
    SYSTEM = "system"
    CUSTOM = "custom"


class CommandStatus(str, Enum):
    """Status of a command in the execution pipeline."""

    PENDING = "pending"
    QUEUED = "queued"
    EXECUTING = "executing"
    COMPLETED = "completed"
    FAILED = "failed"
    TIMEOUT = "timeout"
    CANCELLED = "cancelled"


class CommandSource(str, Enum):
    """Source of the command request."""

    WEB = "web"
    API = "api"
    MCP = "mcp"
    AUTOMATION = "automation"


class CommandExecutionMethod(str, Enum):
    """How a command was or will be executed."""

    DIRECT = "direct"
    POLL = "poll"


class ServiceAction(str, Enum):
    """Actions that can be performed on services."""

    START = "start"
    STOP = "stop"
    RESTART = "restart"
    RELOAD = "reload"


class CommandTarget(BaseModel):
    """Target for a command."""

    node_id: Annotated[str, Field(alias="nodeId", description="Target node ID")]
    service_id: Annotated[
        str | None,
        Field(default=None, alias="serviceId", description="Target service ID (for service commands)"),
    ]

    model_config = {"populate_by_name": True}

    @field_validator("node_id")
    @classmethod
    def validate_node_id(cls, v: str) -> str:
        if not validate_node_id_strict(v):
            raise ValueError(f"Invalid node ID format. Must match pattern: {NODE_ID_PATTERN_NEW}")
        return v

    @field_validator("service_id")
    @classmethod
    def validate_service_id(cls, v: str | None) -> str | None:
        if v is None:
            return v
        if not validate_service_id(v):
            raise ValueError(f"Invalid service ID format. Must match pattern: {SERVICE_ID_PATTERN}")
        return v


class CommandResult(BaseModel):
    """Result of command execution."""

    success: bool = Field(description="Whether the command succeeded")
    output: str | None = Field(default=None, description="Command output")
    exit_code: Annotated[int | None, Field(default=None, alias="exitCode", description="Exit code")]
    error: str | None = Field(default=None, description="Error message if failed")

    model_config = {"populate_by_name": True}


class RequestedBy(BaseModel):
    """Information about who requested the command."""

    user_id: Annotated[str | None, Field(default=None, alias="userId")]
    source: CommandSource = Field(default=CommandSource.API)

    model_config = {"populate_by_name": True}


class CreateCommandRequest(BaseModel):
    """Request to create/queue a new command."""

    type: CommandType = Field(description="Type of command")
    target: CommandTarget = Field(description="Target node/service")
    action: str = Field(description="Action to execute")
    parameters: dict | None = Field(default=None, description="Action parameters")
    timeout_seconds: Annotated[
        int,
        Field(
            default=60,
            ge=1,
            le=3600,
            alias="timeoutSeconds",
            description="Command timeout in seconds",
        ),
    ]

    model_config = {"populate_by_name": True}


class SubmitCommandResultRequest(BaseModel):
    """Request from agent to submit command execution result."""

    success: bool = Field(description="Whether command succeeded")
    output: str | None = Field(default=None, description="Command output")
    exit_code: Annotated[int | None, Field(default=None, alias="exitCode")]
    error: str | None = Field(default=None, description="Error message if failed")

    model_config = {"populate_by_name": True}


class CommandSummary(BaseModel):
    """Summary view of a command."""

    command_id: Annotated[str, Field(alias="commandId")]
    type: CommandType
    target: CommandTarget
    action: str
    status: CommandStatus
    created_at: Annotated[datetime, Field(alias="createdAt")]

    model_config = {"populate_by_name": True}


class CommandResponse(BaseModel):
    """Full command response."""

    command_id: Annotated[str, Field(alias="commandId")]
    type: CommandType
    target: CommandTarget
    action: str
    parameters: dict | None = None
    status: CommandStatus
    result: CommandResult | None = None
    requested_by: Annotated[RequestedBy | None, Field(default=None, alias="requestedBy")]
    timeout_seconds: Annotated[int, Field(alias="timeoutSeconds")]
    created_at: Annotated[datetime, Field(alias="createdAt")]
    queued_at: Annotated[datetime | None, Field(default=None, alias="queuedAt")]
    started_at: Annotated[datetime | None, Field(default=None, alias="startedAt")]
    completed_at: Annotated[datetime | None, Field(default=None, alias="completedAt")]

    model_config = {"populate_by_name": True}


class CommandQueuedResponse(BaseModel):
    """Response when a command is queued or executed directly."""

    command_id: Annotated[str, Field(alias="commandId")]
    type: CommandType
    target: CommandTarget
    action: str
    status: CommandStatus = CommandStatus.QUEUED
    execution_method: Annotated[
        CommandExecutionMethod,
        Field(default=CommandExecutionMethod.POLL, alias="executionMethod"),
    ]
    result: CommandResult | None = None
    queued_at: Annotated[datetime | None, Field(default=None, alias="queuedAt")]
    completed_at: Annotated[datetime | None, Field(default=None, alias="completedAt")]

    model_config = {"populate_by_name": True}


class CommandCancelledResponse(BaseModel):
    """Response when a command is cancelled."""

    command_id: Annotated[str, Field(alias="commandId")]
    status: CommandStatus = CommandStatus.CANCELLED
    cancelled_at: Annotated[datetime, Field(alias="cancelledAt")]

    model_config = {"populate_by_name": True}


class PolledCommand(BaseModel):
    """Typed command payload returned to polling agents."""

    command_id: Annotated[str, Field(alias="commandId")]
    type: CommandType
    action: str
    target: CommandTarget
    parameters: dict | None = None
    timeout_seconds: Annotated[int, Field(alias="timeoutSeconds")]

    model_config = {"populate_by_name": True}


class CommandPollResponse(BaseModel):
    """Response for agent command polling."""

    commands: list[PolledCommand] = Field(
        default_factory=list,
        description="Pending commands for the node",
    )


class CommandResultSubmittedResponse(BaseModel):
    """Response when command result is submitted."""

    command_id: Annotated[str, Field(alias="commandId")]
    status: CommandStatus
    completed_at: Annotated[datetime, Field(alias="completedAt")]

    model_config = {"populate_by_name": True}


class CommandListParams(BaseModel):
    """Parameters for listing commands."""

    node_id: str | None = Field(default=None, alias="nodeId")
    type: CommandType | None = None
    status: CommandStatus | None = None
    since: datetime | None = None
    limit: int = Field(default=50, ge=1, le=200)
    offset: int = Field(default=0, ge=0)

    model_config = {"populate_by_name": True}
